applications/OptimizationApplication/python_scripts/controls/control.py

```diff
@@ -0,0 +1,114 @@
+from abc import ABC, abstractmethod
+from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import SupportedSensitivityFieldVariableTypes
+from KratosMultiphysics.OptimizationApplication.utilities.union_utilities import ContainerExpressionTypes
+
+class Control(ABC):
+    """Base abstract control class.
+
+    The base abstract control class has the following responsibilities.
+        1. Initalizes and finalizes the control.
+        2. Mapping physical space gradients to control space gradients.
+        2. Updating the controlled parts of the model part with new design given in control space.
+        3. Retrieve control field.
+        4. Retrieve physical space kratos variables (If more than one physical space kratos variables are controlled by the given control field.)
+
+    This control should only work on one model part and one kratos control variable. Hence, if multiple model parts required then,
+    a single model part should be created using Kratos.ModelPartOperationUtilities.
+
+    """
+    @abstractmethod
+    def Initialize(self) -> None:
+        """Initializes the control.
+
+        This method initializes the control. This is only called once in the whole optimization process.
+
+        """
+        pass
+
+    @abstractmethod
+    def Check(self) -> None:
+        """Checks the control.
+
+        This method checks the control. This is only called once in the whole optimization process.
+
+        """
+        pass
+
+    @abstractmethod
+    def Finalize(self) -> None:
+        """Finalizes the control.
+
+        This method finalizes the control. This is only called once in the whole optimization process.
+
+        """
+        pass
+
+    @abstractmethod
+    def GetPhysicalKratosVariables(self) -> 'list[SupportedSensitivityFieldVariableTypes]':
+        """Returns list of physical control variables controlled by this control.
+
+        This method returns list of physical variable which are controlled by the control. In most of the cases, there is only one
+        control variable per control. But in some rare cases, there may be the requirement to control two physical variables. Anyhow,
+        the control space should only have one variable per control in the control space, but they can have multiple physical variables
+        in their design space.
+
+        If the case is that this control has one physical variable, then it should be same as the control variable.
+
+        Returns:
+            list[SupportedSensitivityFieldVariableTypes]: List of physical control variables.
+        """
+        pass
+
+    @abstractmethod
+    def GetEmptyControlField(self) -> ContainerExpressionTypes:
```

Comment:
Also I would not make it abstract since it is shared between all controls. Also I would make self.control_model_part at the base class

Reply:
"This is weird to get an empty ContainerExpression to get control's model parts." Please explain what you mean as get control's model part. I am not sure if I get you correctly.

Reply:
I would say it is not weird and it is required. Returning only a model part is not enough, because model parts has, nodes/conditions/elements and control needs to provide which container of model part it is controlling. This empty control field has all the information.

I would keep the base control class fully clean and abstract and not have any member variables so users can call through predefined member functions to retrieve data.

Reply:
renamed to `GetEmptyField`